stop: prefer launching goroutines at caller

tbg opened the pull request:
Prior to this change, the only option to launch async tasks was the stopper's
`RunAsyncTask{,Ex}` method. This method would accept a closure and invoke it in
a goroutine.

This has the undesirable property of recording a location in
`Stopper.RunAsyncTaskEx` as the async task's creating goroutine plus its
lowermost call frame. Anyone who's ever looked at any profile (CPU, heap,
anything really) will be familiar with this fact.

This is a particular problem for Go execution traces, since the root call frame
of the goroutine becomes its searchable identity. In other words, it becomes an
ordeal to search for a specific goroutine, since all goroutines started by the
Stopper get assigned the same searchable identity (root stack frame).

Exec trace regions or tasks don't solve this problem because execution traces
only record those that are set _while an execution trace is taken_. We have
numerous long-lived goroutines in the system.

----

We address these difficulties by introducing a new API that obligates the
*callers* to launch the goroutines directly. In doing so, we also shave off all
unnecessary allocations (from three down to just one, for actually launching
the goroutine).

The API is hopefully self-explanatory and reasonably economical - see
`example_test.go`.

The "old way" to start a task via `StartAsyncTaskEx` remains in place. The
hope is that we'll convert all important tasks to `GetHandle`. This can
be done as follow-up work.

----

```
$ go test -run - -bench BenchmarkStopper -v -test.benchtime=3s ./pkg/util/stop/
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/stop
cpu: Apple M1 Pro
BenchmarkStopper
BenchmarkStopper/RunTask
BenchmarkStopper/RunTask-10         	140792008	        25.48 ns/op	       0 B/op	       0 allocs/op
BenchmarkStopper/AsyncTaskEx
BenchmarkStopper/AsyncTaskEx-10     	 3096800	      1169 ns/op	      72 B/op	       3 allocs/op
BenchmarkStopper/Handle
BenchmarkStopper/Handle-10          	 3165979	      1173 ns/op	      48 B/op	       1 allocs/op
PASS
ok  	github.com/cockroachdb/cockroach/pkg/util/stop	16.834s
```

Epic: CRDB-42584
Release note: None

tbg committed:
stop: prefer launching goroutines at caller

Prior to this change, the only option to launch async tasks was the stopper's
`RunAsyncTask{,Ex}` method. This method would accept a closure and invoke it in
a goroutine.

This has the undesirable property of recording a location in
`Stopper.RunAsyncTaskEx` as the async task's creating goroutine plus its
lowermost call frame. Anyone who's ever looked at any profile (CPU, heap,
anything really) will be familiar with this fact.

This is a particular problem for Go execution traces, since the root call frame
of the goroutine becomes its searchable identity. In other words, it becomes an
ordeal to search for a specific goroutine, since all goroutines started by the
Stopper get assigned the same searchable identity (root stack frame).

Exec trace regions or tasks don't solve this problem because execution traces
only record those that are set _while an execution trace is taken_. We have
numerous long-lived goroutines in the system.

----

We address these difficulties by introducing a new API that obligates the
*callers* to launch the goroutines directly. In doing so, we also shave off all
unnecessary allocations (from three down to just one, for actually launching
the goroutine).

The API is hopefully self-explanatory and reasonably economical - see
`example_test.go`.

The "old way" to start a task via `StartAsyncTaskEx` remains in place. The
hope is that we'll convert all important tasks to `GetHandle`. This can
be done as follow-up work.

----

```
$ go test -run - -bench BenchmarkStopper -v -test.benchtime=3s ./pkg/util/stop/
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/stop
cpu: Apple M1 Pro
BenchmarkStopper
BenchmarkStopper/RunTask
BenchmarkStopper/RunTask-10         	140792008	        25.48 ns/op	       0 B/op	       0 allocs/op
BenchmarkStopper/AsyncTaskEx
BenchmarkStopper/AsyncTaskEx-10     	 3096800	      1169 ns/op	      72 B/op	       3 allocs/op
BenchmarkStopper/Handle
BenchmarkStopper/Handle-10          	 3165979	      1173 ns/op	      48 B/op	       1 allocs/op
PASS
ok  	github.com/cockroachdb/cockroach/pkg/util/stop	16.834s
```

Epic: CRDB-42584
Release note: None

kvoli reviewed (commented):
Left a question. Noticed there were some span after use related test failures, do you know whats going on there?

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-:-OLu_mKV49Hi6afEP6-a:bic6yor)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/util/stop/handle.go` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OLuYz3t1gy-DS_-Sevt:-OLuYz3t1gy-DS_-Sevu:b-pftq8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e353f0bc2a33b633be7c7a4caabb79ad63436e96/pkg/util/stop/handle.go#L61)):*
> ```Go
> 	hdl.s.recover(ctx)
> 	hdl.region.End()
> 	hdl.sp.Finish()
> ```

Should `hld.sp.Finish()` go before `hdl.region.End()` here (Just based off the ordering before in `RunTaskAsyncEx`)?
___
*[`pkg/util/stop/stopper.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OLtGqhj-QickZWwKCNN:-OLtGqhj-QickZWwKCNO:be1m96v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d68fc37181b9e58ecba8c683721fd2a42d54bca4/pkg/util/stop/stopper.go#L438)):*
> ```Go
> 	go func(ctx context.Context) {
> 		ctx, release := hdl.Activate(ctx)
> 		defer release(ctx, hdl)
> ```

minor nit: could put this in the first commit, or were you intending to squash these?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Left a question. Noticed there were some span after use related test failures, do you know whats going on there?

Yeah, I think I have an idea. Previously, the stopper created the trace span on the caller's goroutine, and would then hand it over to the goroutine it spawned (calling `sp.SetGoroutineIDToCurrent()`). There was an ominous comment about "we can't create the span on the goroutine itself because the parent may at that time have finished".

I didn't think our code would so aggressively enforce that ordering, seems a little scary. I was fine with occasionally "missing" the link to the parent span (if the parent span didn't bother waiting around anyway), but crashing is no good.

I'll have to look into the failures and think about what can be done. With the proposed API, on which I've already iterated quite a bit, it is awkward to create the span early. If you do, then you also need to return a Context alongside the Handle (because the caller might reasonably expect to be passing the context that has the span to the goroutine, and in fact may want to wrap it in an additional cancel, or whatever). I'll figure something out.

Before I do though, I'd like to hear if you're generally on board with/in favor of pursuing this PR. I feel fairy strongly that it'll serve us much better (because preserves the relationship between caller and callee better), but if there are concerns let's talk through them first.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-:-OLucCNLAGy4Ye7jguaM:b2bmqf4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: adopt GetHandle in raftScheduler

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142059)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/util/stop/handle.go` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OLuYz3t1gy-DS_-Sevt:-ONOZWla1lD955yLTX28:b-a5qvvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e353f0bc2a33b633be7c7a4caabb79ad63436e96/pkg/util/stop/handle.go#L61)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Should `hld.sp.Finish()` go before `hdl.region.End()` here (Just based off the ordering before in `RunTaskAsyncEx`)?
</blockquote></details>

Done - span is now finished last, since it's created first.


<!-- Sent from Reviewable.io -->

tbg commented:
I cleaned this up, so it's ready again.

tbg committed:
stop: add TestHandleWithoutActivateOrRelease

This shows that expected mistakes lead to a stuck drain
and should be caught quickly.

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 7 files at r5, 3 of 3 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-:-OP0o2MyA56dUoFRQ7Qn:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/util/stop/stopper_test.go` line 730 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OP0oMh-1uG3lstEoDbF:-OP0oMh08U_PjodIjQ3p:b-a2znj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/afdfea3090986ddf32dbb7562a04b6d619c2cd57/pkg/util/stop/stopper_test.go#L730)):*
> ```Go
> 			close(stopCh)
> 		}()
> 		select {
> ```

nit: you could modify this test to use `require.Never` in conjunction with an `var atomic.Bool reached` that's set in the async function above for readability. 
___
*[`pkg/util/stop/handle.go` line 28 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OP0lCDEEooSBasx7OJE:-OP0lCDEEooSBasx7OJF:b34rogj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c056b5964ffe11f73d24c4ce012b07dd8aa25083/pkg/util/stop/handle.go#L28)):*
> ```Go
> 	sp       *tracing.Span       // possibly nil (but nil is functional)
> 
> 	// The below fields are allocated only in Activate, i.e. on the async
> ```

nit: s/below fields/fields below/g
___
*[`pkg/util/stop/handle.go` line 41 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OP0mIUH-tqr-5kdn-Z2:-OP0mIUH-tqr-5kdn-Z3:bb741sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c056b5964ffe11f73d24c4ce012b07dd8aa25083/pkg/util/stop/handle.go#L41)):*
> ```Go
> 
> func (hdl *Handle) Activate(ctx context.Context) ActiveHandle {
> 	growstack.Grow()
> ```

TIL!
___
*[`pkg/util/stop/handle.go` line 44 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OP0n-zG80lZzEn2gd93:-OP0n-zG80lZzEn2gd94:b-okux9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c056b5964ffe11f73d24c4ce012b07dd8aa25083/pkg/util/stop/handle.go#L44)):*
> ```Go
> 
> 	hdl.region = hdl.s.startRegion(ctx, hdl.taskName)
> 	// NB: it's tempting for ergonomics to make `release` a method on `Handle` and
> ```

nit: consider adding a comment on `ActiveHandle` and moving this bit of explanation on there. As is, it's not obvious why we need the `ActiveHandle` interface until you read this tucked away comment.


<!-- Sent from Reviewable.io -->

tbg committed:
stop: improve comments

tbg reviewed (commented):
> Dismissed @arulajmani and @kvoli from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142059)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/util/stop/stopper_test.go` line 730 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/142059#-OP0oMh-1uG3lstEoDbF:-OQ98IqX2gEBcdyD6fPF:brg2g1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/afdfea3090986ddf32dbb7562a04b6d619c2cd57/pkg/util/stop/stopper_test.go#L730)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you could modify this test to use `require.Never` in conjunction with an `var atomic.Bool reached` that's set in the async function above for readability.
</blockquote></details>

Hmm, maybe call me old fashioned, but I'll prefer what's here, despite (with all the time that's passed) barely recognizing myself as the author. I'll leave as is as you don't seem to feel strongly.


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r+
